add c10d dynamic loading mechanism and unit test

ftian1 opened the pull request:
The original behavior of pytorch c10d only supports built-in c10d backends, such as
nccl/gloo/mpi. This patch is used to extend the c10d capability to support dynamically
loading 3rd party communication libraries which are derived from ProcessGroup base class.

related RFC is in: https://github.com/pytorch/pytorch/issues/27955

Through this way, user just need specify a 3rd party c10d backend name when invoking
torch.distributed.init_process_group(). The proposed logic will try to load corresponding
c10d backend cpp extension automatically. as for how to develop a new 3rd party c10d backend
through cpp extension, pls refer to test/cpp_extensions/cpp_c10d_extension.cpp

mrshenli commented on test/test_c10d.py:
```diff
@@ -3205,6 +3206,32 @@ def test_broadcast_coalesced_gloo_cpu(self):
+            process_group = c10d.distributed_c10d.ProcessGroupC10DTEST(store, 0, 1, '')
+
+            work = process_group.allreduce(torch.rand(1))
+            c10d.destroy_process_group()
```
IIUC, this destroys the default group, but `c10d.distributed_c10d.ProcessGroupC10DTEST(store, 0, 1, '')` would create a new instance that's different from the default one?

ftian1 replied:
yes, you are right, I update the logic to use existing process group.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,20 @@ def __new__(cls, name):
                              "on CPU tensors.")
         elif value == Backend.UNDEFINED:
-            raise ValueError("Invalid backend: '{}'".format(name))
+            try:
```
Can we add some docs to this method to explain the the new feature?

ftian1 replied:
I add it in latest commit, pls help check it out

mrshenli reviewed (commented):
Thanks a lot for adding this!!! This will be very useful for other backends as well.

mshiryaev commented:
@ftian1, could you please also include C10D related headers into PyTorch installation to have ability to build external ProcessGroup which depends from these headers?

diff --git a/setup.py b/setup.py
index 7ada277..68c6cb1 100644
--- a/setup.py
+++ b/setup.py
@@ -813,6 +813,7 @@ if __name__ == '__main__':
                 'include/c10/cuda/impl/*.h',
                 'include/c10/hip/*.h',
                 'include/c10/hip/impl/*.h',
+                'include/c10d/*.hpp',
                 'include/caffe2/**/*.h',
                 'include/torch/*.h',
                 'include/torch/csrc/*.h',

ftian1 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -493,7 +509,14 @@ def _new_process_group_helper(world_size,
+                rank,
+                world_size,
+                group_name)
+            _pg_map[pg] = (pg_name, store)
```
yes, it's an assumption that module name 'xyz' can be used to infer usable interface ‘ProcessGroupXYZ’. it's because we have to have a rule to predict such interface symbol...

jgong5 replied:
@ftian1 I think what @mrshenli meant was that whether it is fine to use `("xyz", store)` instead of `("ProcessGroupXYZ", store)` in `_pg_map[pg]` since other backends use the former naming convention. You can still use `globals()["ProcessGroupXYZ"]` to get `pg` here.

ftian1 commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,97 @@
+
+namespace c10d {
+
+std::shared_ptr<ProcessGroup> ProcessGroupC10dTest::createProcessGroupC10dTest(const std::shared_ptr<::c10d::Store>& store, int rank, int size, const std::string& groupName = "") {
```
done

ftian1 commented:
@mrshenli @mshiryaev pls help to check the patch again.

jgong5 commented on docs/source/distributed.rst:
```diff
@@ -10,35 +10,38 @@ Distributed communication package - torch.distributed
+
+
++------------+-----------+-----------+-----------+-----------+
+| Backend    | ``gloo``  | ``mpi``   | ``nccl``  | ``others``|
```
Adding "others" column here doesn't seem to add more useful info. How about keep the original table as is and just mention the usage with text?

ftian1 replied:
fixed

jgong5 commented on docs/source/distributed.rst:
```diff
@@ -395,6 +399,20 @@ of 16
 .. autofunction:: all_gather_multigpu
 
 
+third-party backends
```
"Third-party"

ftian1 replied:
fixed

jgong5 commented on docs/source/distributed.rst:
```diff
@@ -395,6 +399,20 @@ of 16
+please refer to `Tutorials - Custom C++ and CUDA Extensions <https://pytorch.org/
+tutorials/advanced/cpp_extension.html>`_
+
+The backend is derived from `c10d.ProcessGroup` base class like what GLOO/NCCL/MPI made.
```
... like how GLOO/NCCL/MPI backends are made.

ftian1 replied:
done

jgong5 commented on docs/source/distributed.rst:
```diff
@@ -395,6 +399,20 @@ of 16
+The derived class should be named as `ProcessGroupXYZ` if the backend module name is `xyz`. 
+When user invokes :func:`torch.distributed.init_process_group` with a specified backend
+name such as `xyz`, the `torch.distributed` package attempts to import the module `xyz`
+automatically as well as it installed.
```
Shall we add a link to the test case you added as an example? Alternatively, you may also link to how C10D is extended with CCL if the code is ready.

ftian1 replied:
done.

mshiryaev commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,101 @@
+    std::vector<std::vector<at::Tensor>>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllgatherOptions& opts) {
+  throw std::runtime_error("ProcessGroupC10dTest does not support allgather");
```
Wonder why we throw exception in case of allgather but "return NULL" in case of allreduce?

ftian1 replied:
I updated the logic to return a dummy work for only broadcast() and allreduce(). others are still unsupported

mshiryaev commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,90 @@
+  explicit ProcessGroupC10dTest(int rank = -1, int size = -1);
+  virtual ~ProcessGroupC10dTest();
+
+  std::shared_ptr<ProcessGroup::Work> broadcast(
```
Just style-related comment - I would prefer to have the same ordering for functions in header and cpp files just to be aligned.

ftian1 replied:
sovled.

mshiryaev commented on test/test_c10d.py:
```diff
@@ -3205,6 +3206,31 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            store = c10d.FileStore(file.name, 1)
+            process_group = c10d.distributed_c10d._get_default_group()
+            work = process_group.allreduce(torch.rand(1))
```
do we need to add some assert for this block of test?

ftian1 replied:
done by adding a dumy work and wait work success

ftian1 commented on test/test_c10d.py:
```diff
@@ -3205,6 +3206,31 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        self.assertEqual(c10d.get_world_size(), 1)
+
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            store = c10d.FileStore(file.name, 1)
```
it's a typo, should be removed

ftian1 commented:
@mshiryaev @jgong5 @mrshenli I updated the patch according to you guys comments. thanks.

mrshenli reviewed (commented):
Test failures are real:

```
Oct 18 04:57:19 test_broadcast_coalesced_gloo_cuda (__main__.CommTest) ... THCudaCheck FAIL file=/var/lib/jenkins/workspace/aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
Oct 18 04:57:19 THCudaCheck FAIL file=/var/lib/jenkins/workspace/aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
Oct 18 04:57:19 ERROR:root:Caught exception: 
Oct 18 04:57:19 Traceback (most recent call last):
Oct 18 04:57:19   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 133, in wrapper
Oct 18 04:57:19     fn(self)
Oct 18 04:57:19   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 35, in wrapper
Oct 18 04:57:19     return func(*args, **kwargs)
Oct 18 04:57:19   File "test_c10d.py", line 3198, in test_broadcast_coalesced_gloo_cuda
Oct 18 04:57:19     self._test_broadcast_coalesced(process_group, device)
Oct 18 04:57:19   File "test_c10d.py", line 3161, in _test_broadcast_coalesced
Oct 18 04:57:19     target = torch.arange(60, dtype=half, device=device).chunk(5)
Oct 18 04:57:19   File "/opt/conda/lib/python3.6/site-packages/torch/cuda/__init__.py", line 198, in _lazy_init
Oct 18 04:57:19     torch._C._cuda_init()
Oct 18 04:57:19 RuntimeError: cuda runtime error (3) : initialization error at /var/lib/jenkins/workspace/aten/src/THC/THCGeneral.cpp:54
```

I asked @gchanan offline regarding the above error. One potential cause is that the new cpp_extension code did not properly initialize cuda or triggered some error, which could lead to errors for subsequent tests. 

```
Oct 18 04:57:19 test_collective_apis (__main__.BackendLoadingTest) ... ok
Oct 18 04:57:19 test_broadcast_coalesced_gloo_cpu (__main__.CommTest) ... ok
Oct 18 04:57:19 test_broadcast_coalesced_gloo_cuda (__main__.CommTest) ... THCudaCheck FAIL file=/var/lib/jenkins/workspace/aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
```

To debug this, setting CUDA_LAUNCH_BLOCKING and `cudaGetLastError()` might help.

mrshenli commented:
@yf225 do you know what is the appropriate way to initialize cuda in cpp extensions?

ftian1 commented:
@mrshenli @yf225 
I reproduced the "RuntimeError: cuda runtime error (3) : initialization error" error. and in fact it can be triggered by adding "from torch.utils.cpp_extension import load" at original test_c10d.py without any my changes. It's very weird and solutions queried from internet is not helpful...  

I will continue to look into it, but if you guys have suggestions I would be appreciate for that.

yf225 commented:
@ftian1 I feel that it might be worthwhile to look at the docs for `torch.utils.cpp_extension.load` https://github.com/pytorch/pytorch/blob/be3808d3b1164a0073f85fc0a58f2c7c27b2ae40/torch/utils/cpp_extension.py#L571-L650 regarding CUDA-related usage, and also examples in `test/test_cpp_extensions.py` regarding CUDA-related compilations.

ftian1 commented:
@mrshenli @yf225
I followed your suggest and dug into why test/test_cpp_extensions.py is ok on "from cpp_extension import load" but test/test_c10d.py is not ok if we add same line to test_c10d.py. 

the root cause is because classes in test_c10d.py is derived from MultiProcessTestCase and invoke self._fork_processes() to start new processes. according to the search result in web, we have to use spawn way rather fork to generate new process and avoid cuda_init() error. But the embarrass thing is   _spawn_processes() only works with python3. python 2 will assert at _spawn_processes() of test/common_distributed.py. 

could you give some comments on this? shall we update all fork() to spawn() and only allow py3 to trigger c10d test by adding it to py2 blacklist?

ddkalamk commented:
@ftian1 Could you please also add a new init_method that explicitly allow backend to discover rank and world_size internally inside custom backend, something similar to what MPI backend does? I propose adding a new "auto" init_method that allows this like I am doing here in https://github.com/ddkalamk/pytorch/commit/84fddd8bc78ccbc013de2e605cf7d5d7763c632f

kostmo commented:
## CircleCI build failures summary
As of commit 420758c:
* 0/4 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=420758c3d76c6d00ff198a216d1241688533a3bd).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ftian1 commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,102 @@
+#include <c10d/ProcessGroup.hpp>
+#include <c10d/Types.hpp>
+#include <c10d/Utils.hpp>
+#include <c10d/Store.hpp>
```
done

ftian1 commented:
@mrshenli Thanks for @pietern suggestions, I modified the code to make "import cpp_extension" dely happens on a separate test process. it works on my local. The latest preci failures, such as "/temp/test_c10d file existed" or "nija not exist", look like test env issue rather my code bug. could you pls help review again? sorry for the so long PR thread...

mrshenli commented:
Given the following failure, it seems that the load has to be called in the parent process, otherwise all children processes will try to create the same file. If the problem is spawn vs fork, could you please try moving the `test_collective_apis` test to `test_c10d_spawn.py`? 

```
Nov 01 10:52:23 ERROR:root:Caught exception: 
Nov 01 10:52:23 Traceback (most recent call last):
Nov 01 10:52:23   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 133, in wrapper
Nov 01 10:52:23     fn(self)
Nov 01 10:52:23   File "test_c10d.py", line 3257, in test_collective_apis
Nov 01 10:52:23     self._load_c10d_test_backend()
Nov 01 10:52:23   File "test_c10d.py", line 3253, in _load_c10d_test_backend
Nov 01 10:52:23     sources=[src],
Nov 01 10:52:23   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 658, in load
Nov 01 10:52:23     build_directory or _get_build_directory(name, verbose),
Nov 01 10:52:23   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1015, in _get_build_directory
Nov 01 10:52:23     os.makedirs(build_directory)
Nov 01 10:52:23   File "/opt/conda/lib/python3.6/os.py", line 220, in makedirs
Nov 01 10:52:23     mkdir(name, mode)
Nov 01 10:52:23 FileExistsError: [Errno 17] File exists: '/tmp/torch_extensions/c10dtest'
```

mrshenli commented:
@yf225 

I wonder if the the above failure is something that we need to fix. Do we actually need to some uuid/temp dir/name for extensions loading? Otherwise, it seems would fail if multiple process try to load the same module?

mshiryaev commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,20 @@ def __new__(cls, name):
-            raise ValueError("Invalid backend: '{}'".format(name))
+            try:
+                if name not in sys.modules:
+                    third_lib = importlib.import_module(name)
```
This will import Python module with the same name as backend name. Just like example, if backend=gloo then it will try to import module "gloo". Such module names will be too abstract for PyTorch C10D-specific transport bindings. What is about to define the explicit rule for naming of such modules, for example "pytorch-<transport>" and try to import here "pytorch-" + name?

@mrshenli, @pietern, @apaszke  - what do you think?

pietern replied:
Good idea. I recommend improving the error checking here. It is possible that other modules with similar names exist and don't export the `ProcessGroupFoo` class. If that's the case, the backend will also be invalid, but likely results in some kind of `InvalidValue` error here instead.

pietern commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+  m.def("ProcessGroupC10DTEST", &ProcessGroupC10dTest::createProcessGroupC10dTest);
```
Can you name this just `ProcessGroupTest`? The `c10d` part is implied.

pietern commented on docs/source/distributed.rst:
```diff
@@ -10,10 +10,14 @@ Distributed communication package - torch.distributed
-``torch.distributed`` supports three backends, each with
-different capabilities. The table below shows which functions are available
+``torch.distributed`` supports three built-in backends, each with different
+capabilities. It also supports third-party c10d backend extensions by dynamic
```
"third-party backends via a dynamic loading mechanism."

pietern replied:
Let's move the mention of third party backends to a separate section below the introduction and table.

pietern commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,102 @@
+namespace c10d {
+
+//
+// ProcessGroupC10dTest implements a dummy bindings for c10d.
```
"implements dummy bindings"

pietern commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,102 @@
+      std::vector<at::Tensor>& tensor,
+      int tag);
+
+  // Creating a new ProcessGroupC10dTest, will initiialize C10dTest if not initialized
```
What is `C10dTest`? I don't understand this comment.

yf225 commented:
> Given the following failure, it seems that the load has to be called in the parent process, otherwise all children processes will try to create the same file. If the problem is spawn vs fork, could you please try moving the `test_collective_apis` test to `test_c10d_spawn.py`?
> 
> ```
> Nov 01 10:52:23 ERROR:root:Caught exception: 
> Nov 01 10:52:23 Traceback (most recent call last):
> Nov 01 10:52:23   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 133, in wrapper
> Nov 01 10:52:23     fn(self)
> Nov 01 10:52:23   File "test_c10d.py", line 3257, in test_collective_apis
> Nov 01 10:52:23     self._load_c10d_test_backend()
> Nov 01 10:52:23   File "test_c10d.py", line 3253, in _load_c10d_test_backend
> Nov 01 10:52:23     sources=[src],
> Nov 01 10:52:23   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 658, in load
> Nov 01 10:52:23     build_directory or _get_build_directory(name, verbose),
> Nov 01 10:52:23   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1015, in _get_build_directory
> Nov 01 10:52:23     os.makedirs(build_directory)
> Nov 01 10:52:23   File "/opt/conda/lib/python3.6/os.py", line 220, in makedirs
> Nov 01 10:52:23     mkdir(name, mode)
> Nov 01 10:52:23 FileExistsError: [Errno 17] File exists: '/tmp/torch_extensions/c10dtest'
> ```

@mrshenli Yes I think we should use temp names for extensions loading. I will file an issue for it after reproducing it with a smaller example.

ftian1 commented:
@mrshenli @yf225 The temp name conflict error for extensions loading has been solved by passing a random dir name. And the ninja missing issue on Multi-GPU test case is workaround by verify ninja availability before run test (why test_cpp_extension has no such issue is because it doesn't run in this test). 

Now the remaining two pre-ci failures seems not be related with my code. could you take a look?

one error is "RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff". another one is "git merge --no-edit --no-ff7073ee209000a7781c0c863c4ef39bb3bfdb4932 fatal: refusing to merge unrelated histories"

Jianhui-Li commented:
@pietern   Looks like the failure is irrelevant, and is this patch ready to be merged in? 

@gottbrath  This is the dynamic loading mechanism which allow us to add intel communication library (named CCL, or previously MLSL) as third-party plug-in, and user needs to explicitly import the capability in the code.  We recently open-sourced CCL https://github.com/intel/oneccl

pietern commented:
@pytorchbot rebase this please

pietern commented on docs/source/distributed.rst:
```diff
@@ -10,8 +10,9 @@ Distributed communication package - torch.distributed
-different capabilities. The table below shows which functions are available
+``torch.distributed`` supports three built-in backends, each with different
+capabilities.
+The table below shows which functions are available on GLOO/MPI/NCCL backends
```
Please revert this, keeping only the "built-in" part. The explicit mention of backends is not needed.

ftian1 replied:
done

pietern commented on docs/source/distributed.rst:
```diff
@@ -50,7 +51,6 @@ MPI is an
 optional backend that can only be included if you build PyTorch from source. (e.g.
 building PyTorch on a host that has MPI installed.)
 
-
```
Why remove this?

ftian1 replied:
a typo, fixed it.

pietern commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
+Third-party backends
+--------------------
+
+Besides GLOO/MPI/NCCL backends, Pytorch distributed supports third-party backends by
```
* "Besides the Gloo/MPI/NCCL..."
* "PyTorch"
* "by" -> "through a"

ftian1 replied:
amended

pietern commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
+
+The backend is derived from `c10d.ProcessGroup` base class like how GLOO/NCCL/MPI backends
+are made.
+The derived class should be named as `ProcessGroupXYZ` if the backend module name is `xyz`. 
```
The new backend derives from `c10d.ProcessGroup` and should be named `ProcessGroupXYZ`...

ftian1 replied:
modified

pietern commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
+The backend is derived from `c10d.ProcessGroup` base class like how GLOO/NCCL/MPI backends
+are made.
+The derived class should be named as `ProcessGroupXYZ` if the backend module name is `xyz`. 
+When user invokes :func:`torch.distributed.init_process_group` with a specified backend
```
When :func:`torch.distributed.init_process_group` is invoked...

ftian1 replied:
modified.

pietern commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
+are made.
+The derived class should be named as `ProcessGroupXYZ` if the backend module name is `xyz`. 
+When user invokes :func:`torch.distributed.init_process_group` with a specified backend
+name such as `xyz`, the `torch.distributed` package attempts to import the module `python_xyz`
```
[...] attempts to import the module `python_xyz`. This module is expected to define the `ProcessGroupXYZ` class.

pietern replied:
See comment below regarding package naming, please update here as well.

ftian1 replied:
modified to reflect current behavior

pietern commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+  m.def("ProcessGroupTEST", &ProcessGroupC10dTest::createProcessGroupC10dTest);
```
Why all caps? Let's define this as `ProcessGroupTest`, both in pybind, as well as the class itself.

pietern commented on test/test_c10d.py:
```diff
@@ -3240,6 +3247,51 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        self.assertTrue(work.is_completed())
+        self.assertTrue(work.is_success())
+
+        c10d.destroy_process_group()
```
This file should only test the c10d API directly. For `torch.distributed` tests, see `test_distributed.py`.

ftian1 replied:
moved the logic to test_distributed.py

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,23 @@ def __new__(cls, name):
+            try:
+                if name not in sys.modules:
+                    # using "pytorch_" prefix to avoid module naming conflicts/confusion 
+                    third_lib = importlib.import_module("pytorch_{}".format(name))
```
Extensions are typically named with a dash separator, e.g. `torch-ccl`.

Also, let's use the `torch` prefix instead of `pytorch`, given that PyTorch itself is published as `torch`.

mshiryaev replied:
@pietern , can "torch" prefix make any confusion for users as it looks like module for old Torch?

pietern replied:
This is the name of the PyTorch package.

Also, it a Python package, so there's little ambiguity :)

mshiryaev replied:
from search in pypi packages for "torch-" and "pytorch-" prefixes I see that both are used for PyTorch modules
if you suggest to use "torch-" then ok, it will make name shorter :)

ftian1 replied:
> Extensions are typically named with a dash separator, e.g. `torch-ccl`.

@pietern , the problem we met is cpp_extension.load() is unable to tolerate "-" character in "name" argument. A similar issue was raised at pytorch/extension-cpp#13. To workaround this, we have to propose "torch_xyz" as extension name rather than "torch-xyz".

when user installs cpp_extension by setup.py, he can specify different extension name and module name to workaound this.

ext_modules = [
    CppExtension(
        name="torch_xyz",
        ...
    )]

setup(
    name='torch-xyz',
    ...)
Through this way, "pip install torch-xyz" and "import torch_xyz" all work.

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,23 @@ def __new__(cls, name):
+                else:
+                    third_lib = sys.modules["pytorch_{}".format(name)]
+
+                pg_name = "ProcessGroup{}".format(name.upper())
```
Upper shouldn't be needed here. We can define a getter function on the module that returns the class instead.

ftian1 replied:
@pietern May I get you clarification on this comment? do you mean we shouldn't assume the exposed instantiation interface name is ProcessGroupXYZ and should get the name from extension?  IMHO anyway we have to have assumption (or naming appointment) on the name unless there is a query interface. but the query interface's name is also an appointment...

Do I misunderstand something?

pietern replied:
No that's my intent indeed.

Following what you implemented now, it will first try to call `name()` on the module to query for the right name. If that doesn't work, it'll try to get the class `ProcessGroupFoo`. This seems like the proper ordering to me.

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
+                elif other_backend.__dict__.get("ProcessGroup{}".format(name.lower().capitalize())):
+                    pg_name = "ProcessGroup{}".format(name.lower().capitalize())
+                else:
+                    raise NameError("Not found '{}' symbol in '{}' python module".format(pg_name, "torch_{}".format(name.lower())))
```
This error will be hidden by the `except NameError` below.

pietern replied:
`pg_name` is undefined when you raise this error.

pietern commented on test/test_distributed.py:
```diff
@@ -27,6 +28,12 @@
+Ninja (https://ninja-build.org) must be available to run C++ extensions tests,
+but it could not be found. Install ninja with `pip install ninja`
+or `conda install ninja`. Alternatively, disable the test with
+`run_test.py --exclude c10d`.
```
This comment is no longer up to date with the exclusion comment.

I think you can remove the "disable" comment because disabling the whole distributed test just for missing ninja is a bit much in my opinion. Having the message show that the test is skipped should be enough.

ftian1 replied:
removed

pietern commented on test/test_distributed.py:
```diff
@@ -1984,6 +1991,7 @@ def setUpClass(cls):
         def setUp(self):
             super(TestDistBackend, self).setUp()
             # We rely on the manager process to delete the temporary file.
+
```
Why add a line here?

ftian1 replied:
removed

pietern commented on test/test_distributed.py:
```diff
@@ -2094,6 +2102,50 @@ def _join_and_reduce(self, fn):
+                torch.utils.cpp_extension.verify_ninja_availability()
+            except RuntimeError:
+                print(CPP_EXTENSIONS_WARNING)
+                return 0
```
You can turn this into a decorator so you can skip the test before running it. Then the test will show up as skipped if ninja is not available. If you return from the test itself, it shows up as successful.

ftian1 replied:
done

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
+                if other_backend.__dict__.get("name"):
+                    pg_name = other_backend.name()
+                elif other_backend.__dict__.get("ProcessGroup{}".format(name.lower().capitalize())):
+                    pg_name = "ProcessGroup{}".format(name.lower().capitalize())
```
You can generate this once, store it in a variable, and use it twice. Copying the name generation is error prone.

ftian1 replied:
done

pietern reviewed (commented):
The dash vs underscore is unfortunate, but if we can't get around it, then we can leave it as is.

Thanks for addressing my comments. I have a couple more, but we're getting close!

pietern commented:
Beware: #29928 was merged and updates the return type of  `Work::wait`.

ftian1 commented:
@pietern many thanks for your review. I fixed them and am hoping it closes to check-in criteria:)

ftian1 committed:
add c10d backend register mechanism to support 3rd party backends.

The original behavior of pytorch c10d only supports built-in backends, such as
nccl/gloo/mpi. This patch is used to extend the c10d capability to support 3rd party
communication libraries which are derived from ProcessGroup base class.

related RFC is in: https://github.com/pytorch/pytorch/issues/27955

Through this way, user just need manually import this backend and specify the backend
name when invoking torch.distributed.init_process_group(). The proposed logic will check
if the backend is registered through torch.distributed.Backend.register_backend().
As for how to develop a new 3rd party backend through cpp extension, pls refer to
test/cpp_extensions/cpp_c10d_extension.cpp

ftian1 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
+                setattr(Backend, name.upper(), name.lower())
+                value = getattr(Backend, name.upper(), Backend.UNDEFINED)
+            except (ImportError, NameError):
+                raise ValueError("Invalid backend: '{}'".format(name))
```
sorry for missing your point.  I have fixed that.

ftian1 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
+
+                globals()[pg_name] = other_backend.__dict__.get(pg_name)
+                setattr(Backend, name.upper(), name.lower())
+                value = getattr(Backend, name.upper(), Backend.UNDEFINED)
```
I updated the code to not use globals()

ftian1 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
+            try:
+                module_name = "torch_{}".format(name.lower())
+                if name not in sys.modules:
+                    # using "torch_" prefix to avoid module naming conflict/confusion 
```
yes. it's my typo:(

ftian1 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -73,7 +76,25 @@ def __new__(cls, name):
-            raise ValueError("Invalid backend: '{}'".format(name))
+            try:
+                module_name = "torch_{}".format(name.lower())
+                if name not in sys.modules:
```
fixed that

pietern reviewed (commented):
@ftian1 

I had a discussion with @fmassa about the `torch_xyz` package pattern yesterday. If we consider the future where you might be trying out multiple backends, or have multiple versions of the same backend (e.g. CCL v1 and CCL v2), the package naming assumptions that we make here will be fragile.

Namely, if you introduce a new backend, you'll have to use that exact package name to make it available to PyTorch. This is pretty strict and can get in the way of easily extending this.

I think it will be better if we remove the package name assumption and the automatic import, and replace it with a call to `Backend.register_new_backend("name", klass)`, which does the same as the code you added to the `Backend` class now, except for the import. Calling this function means the backend is usable without jumping through hoops, we don't need to call `module.name()`, or make assumptions about the name of the ProcessGroup subclass.

For users, this is one extra line of code to import the right package that makes the backend available, which is a small price to pay for better clarity and extensibility in my opinion.

What do you think?

ftian1 commented:
@pietern My original thought was user unawareness through loading backend transparently. Your concern is what I didn't consider. I am ok with this idea, will update it according to your comments

ftian1 commented:
@pietern i updated the patch according to your comments, please help review. thanks:)

mshiryaev commented on torch/distributed/distributed_c10d.py:
```diff
@@ -74,8 +77,14 @@ def __new__(cls, name):
         return value
 
+    @classmethod
+    def register_new_backend(cls, model_name, instantiating_interface_name):
```
> model_name

maybe module_name?

ftian1 replied:
> @ftian1 , @pietern
> Сould you please clarify how new approach will look from user script point of view?
> My current understanding is that user need to add one additional import and correctly specify backend name for init_process_group.
> 
> import torch.nn.parallel
> import torch.distributed as dist
> import <transport>
> 
> ...
> backend = '<transport_name>'
> dist.init_process_group(backend, ...)
> model = torch.nn.parallel.DistributedDataParallel(model, ...)
> ...

yes, you are right. Just pls note the backend paramter passed in dist.init_process_group() should be same with the one registered through Backend.register_new_backend().

and module_name is my typo. I will fix it soon

mshiryaev reviewed (commented):
@ftian1 , @pietern 
Сould you please clarify how new approach will look from user script point of view?
My current understanding is that user need to add one additional import and correctly specify backend name for init_process_group.

import torch.nn.parallel
import torch.distributed as dist
import \<transport\>

...
backend = '<transport_name>'
dist.init_process_group(backend, ...)
model = torch.nn.parallel.DistributedDataParallel(model, ...)
...

pietern commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,115 @@
+      // The first parameter is the python module name, The second parameter is
+      // the instantiating interface name.
+      register_new_backend("torch_test", "createProcessGroupTest");
+  }
```
Looks like a long line -- can you run `clang-format` on this file?

ftian1 replied:
I ran "python tools/clang_format.py -d 190dac13e33267b4808a280fbc449ad672ec7544". no issue reported:). but I agree it's not elegant and update it in code

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -43,7 +45,8 @@
     """
-    An enum-like class of available backends: GLOO, NCCL, and MPI.
+    An enum-like class of available backends: GLOO, NCCL, MPI and other registered
+    backends.
```
Oxford comma after `MPI`.

ftian1 replied:
done

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -74,8 +77,14 @@ def __new__(cls, name):
         return value
 
+    @classmethod
+    def register_new_backend(cls, module_name, instantiating_interface_name):
```
`register_backend` is good enough, `new` is implied.

ftian1 replied:
done

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -74,8 +77,14 @@ def __new__(cls, name):
 
+    @classmethod
+    def register_new_backend(cls, module_name, instantiating_interface_name):
+        setattr(Backend, module_name.upper(), (module_name, instantiating_interface_name))
```
I would change this to register the ProcessGroup subclass directly. This means you still need to look up the module and get the class, whereas this function could just register that class directly.

ftian1 replied:
I think here you meant to register the subclass ProcessGroup instantiation interface directly? I made the change according to this

pietern commented on torch/distributed/distributed_c10d.py:
```diff
@@ -487,7 +496,15 @@ def _new_process_group_helper(world_size,
         else:
-            raise RuntimeError("Unsupported distributed backend by group")
+            module_name, instantiating_interface_name = getattr(Backend, backend.upper())
+            other_backend = sys.modules[module_name]
```
This lookup and the `getattr` below shouldn't be necessary.

ftian1 replied:
The problem is here we couldn't remove getattr. I will explain on the formal reply.

pietern reviewed (commented):
Looking good, thanks for the modifications.

@mrshenli Can you take a look here as well?

mshiryaev commented on torch/distributed/distributed_c10d.py:
```diff
@@ -74,8 +77,14 @@ def __new__(cls, name):
 
+    @classmethod
+    def register_new_backend(cls, module_name, instantiating_interface_name):
+        setattr(Backend, module_name.upper(), (module_name, instantiating_interface_name))
```
What is about to split module name and backend name?

E.g. python module with new transport will have some complex name like "torch_xxx_yyy_zzz" and user will import module by this name, but will pass some short name to init_process_group(backend='aaa', ...).

Otherwise currently user should pass again full module name as backend name.

ftian1 replied:
I update the code to support your idea. now as long as user register a short backend name in register_backend(), then it will be used to match with the one passed in init_process_group().

ftian1 commented:
@pietern @mrshenli I updated the patch according to Pieter's comments except for the last one. Why I think getattr() in _new_process_group_helper() is still needed is because Backend() original semantics is to return string rather class. if we return subclass object on Backend(xxxx) for removing getattr(), it will assert as Backend(xxxx) is invoked many times and assume xxxx must be a string.

To avoid big changes on original code, i keeps getattr() in _new_process_group_helper(). if it's not meet your thoughts, pls let me know:)

mshiryaev commented:
@pietern @mrshenli - could you please review the latest changes from @ftian1? Is there anything that blocks this PR?

pietern commented:
@ftian1 Could you rebase this change to make sure it merges cleanly and passes CI?

pietern commented:
@ftian1 You changed the mode of `distributed_c10d.py` to 0755, please revert this.

ftian1 commented:
@pietern many thanks for your review. now I am rebasing on latest master and trying to get the pre-ci pass.

but some of errors are not because of my PR, such as lint check.
+ cat /home/runner/work/pytorch/pytorch/flake8-output.txt
./torch/jit/__init__.py:1516:25: W605 invalid escape sequence '\s'
./torch/jit/__init__.py:1516:87: W605 invalid escape sequence '\s'

I am trying to reproduce some of errors, it will spend a little time to construct the env. sorry for the inconvenience.

ftian1 commented:
@pietern @mrshenli I have rebased my code to latest without any change. now there remains two pre-ci errors. but it's not related with my code and fails at test_quantization.py. 

May I know how to go ahead?
 
Dec 20 15:23:30 Traceback (most recent call last):
Dec 20 15:23:30   File "test/run_test.py", line 459, in <module>
Dec 20 15:23:30     main()
Dec 20 15:23:30   File "test/run_test.py", line 452, in main
Dec 20 15:23:30     raise RuntimeError(message)
Dec 20 15:23:30 RuntimeError: test_quantization failed!

mrshenli commented:
> now there remains two pre-ci errors. but it's not related with my code and fails at test_quantization.py.

@ftian1 that test failure is irrelevant and also occurs to master. I will leave it to @pietern to re-import the PR for internal tests. Thanks for contributing!

mshiryaev commented:
@pietern - it is just kindly reminder. Seems all your comments were addressed and test failures are not related with changes from this PR.

Jianhui-Li commented:
@pintern Could you please check whether it is ready to merge?  @gottbrath

mrshenli commented:
IIUC, @agolynski 's PR #31892 avoids forcing ProcessGroup backend to implement `allgather_coalesced` (same thing needs to be done for `allreduce_coalesced`) API. 

@agolynski could you please comment here and judge if `allgather_coalesced` and `allreduce_coalesced` need to be removed from this PR?

agolynski commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,119 @@
+
+  std::shared_ptr<ProcessGroup::Work> recvAnysource(
+      std::vector<at::Tensor>& tensor,
+      int tag);
```
do you need to declare these?

ftian1 replied:
yes, it's for build pass.

agolynski commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,122 @@
+    std::vector<at::Tensor>& inputTensors,
+    const AllgatherOptions& opts) {
+  throw std::runtime_error("ProcessGroupTest does not support allgather_coalesced");
+}
```
We (+@mrshenli) have recently made changes to ProcessGroup API and decided that we'll only require implementing allgather_base which gathers just 1 tensor (use one tensor as input and one tensor as output that would contain all gathered tensors concatenated), e.g. see https://github.com/pytorch/pytorch/pull/31892.
However, this process is not completed yet.

Could you implement allgather_base here, otherwise your code would break the build?
Also, we already allow clients not to implement allgather_coalesced, so you can drop this implementation.

Just FYI for now: the same is true for allreduce_coalesced, but we haven't started the process of deprecating it yet.

ftian1 replied:
ok. I will update the code accordingly.

ftian1 commented:
@mrshenli @agolynski I have updated the code to not implement allgather_coalesced but allgather_base. pls check it out

Jianhui-Li commented:
@gottbrath  @pietern could you please approve and merge this PR? it seems everything is ready.

pietern commented:
cc @mrshenli @pritamdamania87

ftian1 commented:
@mrshenli @pritamdamania87 May I know how to make this PR go forward?

gottbrath commented:
@pritamdamania87 can you review and approve this or provide clear guidance on what is needed to have this approved.

ftian1 commented:
@VitalyFedyunin hello, Would you pls help to review the PR?

VitalyFedyunin commented:
I will pass it to @mrshenli as he is most experienced in distributed

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -487,7 +502,13 @@ def _new_process_group_helper(world_size,
+                prefix_store,
+                rank,
+                world_size,
+                group_name)
```
Does this mean all 3rd party backend must use the same signature? What about `timeout`? and why do we need to pass in group_name here?

ftian1 replied:
@mrshenli understand your concern. At my local first version, I added timeout and no group_name, but later removed it as timeout is not necessary for our c10d backend. 

I would add timeout back for all 3rd party backend.

mrshenli reviewed (commented):
I apologize that I dropped the ball on this, and sorry about the long delay on this PR. Let's expedite landing processing for this. 

The current version looks great to me. I have three main comments

1. ideally we should have a well-designed and stable PG interface before we landing this PR. But as I am not sure when that will come, let's don't block landing this PR due to that. If we need to change the PG interface in the future, we will modify this as well. However, if your extension lives out of PyTorch repo, it means there might be conflicts. How bad will that be if we break your extension? @ftian1 

Also, @agolynski please drop notes if you have comments on this.

2. Looks like the following will be THE interface for PG constructor. This might not be sufficient for future PG implementation if they require additional configurations. But we should be able address those by adding sth like `backend_options` later when that requirement emerges. As for now, is there any reason for adding the group_name here? NCCL and GLOO does not require a name but does require a timeout. Shall we at least pass the timeout to the PG as well so that implementations can respect timeout if they would like to?

```python
            pg = getattr(Backend, backend.upper())(
                prefix_store,
                rank,
                world_size,
                group_name)
```

3. Looks like the test require Ninja. Could you please explain why do we need that? 

Hey @yf225, do we have ninja installed in all our CI test envs?

ftian1 commented on test/cpp_extensions/cpp_c10d_extension.cpp:
```diff
@@ -0,0 +1,122 @@
+  return std::make_shared<ProcessGroupTest::WorkTest>();
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::allreduce_coalesced(
```
I will add allreduce_base on the refreshed PR

ddkalamk replied:
What is `allreduce_base` as `allreduce` already expects a just single tensor?

mrshenli replied:
I will leave it to @agolynski to comment on the plan for `allreduce/allgather_base` API. :)

ftian1 commented on test/test_distributed.py:
```diff
@@ -30,6 +31,12 @@
 skipIfNoTorchVision = unittest.skipIf(not HAS_TORCHVISION, "no torchvision")
 
+CPP_EXTENSIONS_WARNING = """
+Ninja (https://ninja-build.org) must be available to run C++ extensions tests,
```
it's because torch c++ extension build needs Ninja. it's copied from test/run_test.py (you can see same thing in run_test.py)

ftian1 commented on test/cpp_extensions/cpp_c10d_extension.hpp:
```diff
@@ -0,0 +1,119 @@
+      const std::string& groupName);
+
+  static void ProcessGroupTestConstructor() __attribute__((constructor)) {
+      py::object module = py::module::import("torch.distributed");
```
according to pybind11 doc, when calling a C++ function from Python, the GIL is always held.

ftian1 commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
+
+When manually importing this backend and invoking :func:`torch.distributed.init_process_group`
+with the corresponding backend name, the `torch.distributed` package runs on the new backend.
+
```
sure.

mrshenli commented:
Hey @agolynski, could you please check if the API proposed in this PR looks OK to you? Thanks!

ftian1 commented:
@mrshenli @agolynski I updated the PR according to your comments, pls let me know if there needs further updates. many thanks for your patients

mrshenli commented on docs/source/distributed.rst:
```diff
@@ -395,6 +395,24 @@ of 16
 
+Third-party backends
+--------------------
+
```
Shall we mark this as experimental for now? You can add sth like:

```
.. warning::
  The third-party backend API is experimental and subject to change.
```

ftian1 replied:
will fix it

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -75,8 +76,22 @@ def __new__(cls, name):
 
+    @classmethod
+    def register_backend(cls, name, func):
+        """Registers a new backend.
```
Does it show properly in the doc? Could you please try the following and paste the screenshot here? Thanks!

```
cd docs/
pip install -r requirements.txt
make html
```

You can check the result by using this [gist](https://gist.github.com/colesbury/df218a96334dbd31ee0e93760f70f114) to serve your docs and paste a screenshot in this PR.

ftian1 replied:
@mrshenli, make html would fail with the below error.

FileNotFoundError: [Errno 2] No such file or directory: 'katex': 'katex'

but lucky thing is distributed.html has been generated. I paste the screenshot for your reference.
https://user-images.githubusercontent.com/16394660/76596980-70036280-653a-11ea-8d57-f36ee54894c2.png

![image](https://user-images.githubusercontent.com/16394660/76596980-70036280-653a-11ea-8d57-f36ee54894c2.png)

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -75,8 +76,22 @@ def __new__(cls, name):
+
+        Arguments:
+            name (str): Backend name matching with the one in `init_process_group()`.
+            func (function): Function handler that instantiates the backend.
```
As we expect the function to take certain arguments (`prefix_store`, `rank`, `world_size`, and `timeout`), can we explain those args here as well? 

Ideally, we might want to make this function first class citizen as well with clear signature and documents. But as we can mark this feature as experimental, we should be able to have some room to change it in the future when necessary.

ftian1 replied:
will add comments here

mrshenli reviewed (commented):
Thanks a lot for putting this together!  Overall, it LGTM! I left some minor comments on documentation. 

@agolynski will comment on the c10d API.

mrshenli reviewed (changes requested):
Test failure is real:


```
Mar 10 01:34:51 Traceback (most recent call last):
Mar 10 01:34:51   File "test_determination.py", line 97, in test_torch_file
Mar 10 01:34:51     "test_determination",
Mar 10 01:34:51 AssertionError: Lists differ: ['distributed/test_distributed', 'test_cpp_[75 chars]ion'] != ['test_cpp_extensions_aot_ninja', 'test_cpp[43 chars]ion']
Mar 10 01:34:51 
```

ftian1 commented:
@mrshenli I refine the doc/code according to your comment. pls let me know if it needs further update. thanks

ddkalamk commented:
@mrshenli @agolynski 

May I request your time to make progress on this PR? As I mentioned earlier, this PR and https://github.com/pytorch/pytorch/pull/32361 are critical for our work with @srinivas212 and @dmudiger on distributed DLRM training on CPUs. 
Also, with small change, this PR would enable building MPI backend to be used with binary distribution of pytorch saving pain to build whole pytorch from sources just to use MPI backend.

mrshenli commented:
Sorry about the delay. @agolynski is working on the last pass of the review and if everything looks good, he will land this soon. 

Thanks for contributing!

esaliya commented:
@ftian1 I am trying out a third-party backend using a custom extension as implemented in this PR. However, new_group() creation does not send the ranks list in the sub-group to the registered method in the extension. Could you suggest how to get this? Here's the function call I am referring to https://github.com/pytorch/pytorch/blob/b9f02778b29474ad47bc4f44d727a922bc469853/torch/distributed/distributed_c10d.py#L724

ftian1 commented:
@esaliya could you pls refer to this example? https://github.com/intel/torch-ccl/blob/master/torch_ccl/csrc/init.cpp

it's a backend we added for this PR

esaliya commented:
Thanks @ftian1. I looked at it but it too doesn't have a way to get the list of `ranks` when creating a `new_group()`. For example,

```
dist.init_process_group('ccl', rank=3, world_size=8)
dist.new_group(ranks=[0,1,2])
```
in the second line, the list of ranks, `[0,1,2]` is passed to this https://pytorch.org/docs/stable/_modules/torch/distributed/distributed_c10d.html#new_group but they are **not** passed to the ccl backend here https://github.com/pytorch/pytorch/blob/b9f02778b29474ad47bc4f44d727a922bc469853/torch/distributed/distributed_c10d.py#L724

Any thoughts on this?

ftian1 commented:
ok, I understand. when I added such dynamic loading mechanism I didn't take such sub_group/group_ranks into count. 

so if your backend requests such capability, you have to contribute a PR to add this

esaliya commented:
Thanks, for confirming this.
